src/shrinkage.jl

```diff
@@ -1,91 +1,46 @@
-        levs = re.levels
-        refs = re.refs
-        grpest = Matrix{Union{Missing,T}}(missing, (length(levs), length(cnms)))
-        for i in eachindex(levs)
-            rows = findall(==(i), refs)
-            try
-                grpest[i, :] = view(Xmat, rows, :) \ view(yvec, rows)
-            catch
-            end
-        end
-        push!(cvec, CoefByGroup(cnms, levs, Xmat \ yvec, ranefs[j]', grpest))
+function shrinkageplot(
+    m::LinearMixedModel{T},
+    gf::Symbol=first(fnames(m)),
+    θref::AbstractVector{T}=10000 .* m.optsum.initial,
+) where {T}
+    reind = findfirst(==(gf), fnames(m))  # convert the symbol gf to an index
+    if isnothing(reind)
+        throw(ArgumentError("gf=$gf is not one of the grouping factor names, $(fnames(m))"))
     end
-    NamedTuple{fnames(m)}((cvec...,))
-end
-
-shrinkageplot(m::LinearMixedModel, gf::Symbol=first(fnames(m))) = shrinkageplot(shrinkage(m)[gf])
-
-shrinkageplot(cg::CoefByGroup) = shrinkageplot!(Figure(resolution=(1000,1000)), cg)
-
-function shrinkageplot!(f::Figure, cg::CoefByGroup)
-    shrinkage2d!(Axis(f[1,1]), cg)
-    f
-end
-
-function shrinkage2d!(a::Axis, cg::CoefByGroup{T}, inds=(1, 2)) where T
-    i, j = inds
-    fxd = cg.fixed
-    scatter!(a, view(fxd, i:i), view(fxd, j:j), color=:green, label="Pop")
-    u = view(cg.condmodes, :, i) .+ fxd[i]
-    v = view(cg.condmodes, :, j) .+ fxd[j]
-    scatter!(a, u, v, color=:blue, label="Mixed")
-    x = view(cg.grpest, :, i)
-    y = view(cg.grpest, :, j)
-    missgrp = ismissing.(x) .| ismissing.(y)
-    if any(missgrp)
-        nonmiss = .!(missgrp)
-        x = convert(Vector{T}, view(x, nonmiss))
-        y = convert(Vector{T}, view(y, nonmiss))
-        u = view(u, nonmiss)
-        v = view(v, nonmiss)
-    else
-        x = convert(Vector{T}, x)
-        y = convert(Vector{T}, y)
+    reest = ranef(m)[reind]               # random effects conditional means at estimated θ
+    reref = ranef(updateL!(setθ!(m, θref)))[reind]  # same at θref
+    updateL!(setθ!(m, m.optsum.final))    # restore parameter estimates and update m
+    cnms = m.reterms[reind].cnames
+    f = Figure(; resolution=(1000, 1000)) # use an aspect ratio of 1 for the whole figure
+    k = size(reest, 1)  # dimension of the random-effects vector per level of gf
+    for i in 2:k                          # strict lower triangle of panels
+        for j in 1:(i - 1)
+            ax = Axis(f[i - 1, j])
+            x, y = view(reref, j, :), view(reref, i, :)
+            scatter!(ax, x, y; color=(:red, 0.25))   # reference points
+            u, v = view(reest, j, :), view(reest, i, :)
+            arrows!(ax, x, y, u .- x, v .- y)        # first so arrow heads don't obscure pts
+            scatter!(ax, u, v; color=(:blue, 0.25))  # conditional means at estimates
+			if i == k              # add x labels on bottom row
+				ax.xlabel = string(cnms[j])
+			else
+				hidexdecorations!(ax; grid=false)
```

Comment:
should there be a call to `linkyaxes!` for each row and `linkxaxes!`  for each column?

Reply:
Probably.  I don't think it is necessary when the range of the data values is the same on those axes but it can't hurt.  Do you want to add that?

A related issue is whether to adjust the spacing between panels.  But that brings us back to the earlier question of overall design philosophy.

Reply:
I added `linkaxes!` calls. Let me know if it's efficient enough for you. :smile:

Reply:
Looks good to me.